### PR TITLE
Fix multithreading. Fix #291 - SSLError

### LINKED ISSUE #291: SSLError: [SSL: DECRYPTION_FAILED_OR_BAD_RECORD_MAC] decryption failed or bad record mac
I have a python script which uses pygsheets to connect to gsheet. The script runs quite well on the Windows Server but throws an exception 'SSLError: [SSL: DECRYPTION_FAILED_OR_BAD_RECORD_MAC] decryption failed or bad record mac' when running on Ubuntu.
`````import pygsheets
from oauth2client.service_account import ServiceAccountCredentials

scope = [
        'https://spreadsheets.google.com/feeds',
        'https://www.googleapis.com/auth/drive'
         ]

credentials01 = ServiceAccountCredentials.from_json_keyfile_name('creds01.json', scope)
credentials02 = ServiceAccountCredentials.from_json_keyfile_name('creds02.json', scope)

file01 = pygsheets.authorize(credentials=credentials01)
file02 = pygsheets.authorize(credentials=credentials02)
wb01 = file01.open('Database System 2')
wb02 = file02.open('Database System 2')

if __name__ == '__main__':
    manager = multiprocessing.Manager()
    output_dfs = manager.list()
    new_dfs_dict = manager.dict()
    for l in main_list:
        for i in range(0, len(l[0]), 5):
            processes = []
            ids_sublist = []
            for j in range(i, i + 5):
                try:
                    ids_sublist.append(l[0][j])
                except:
                    pass
            for ID in ids_sublist:
                processes.append(multiprocessing.Process(target=fetch_data, args=(ID, start_date, end_date, new_dfs_dict, output_dfs)))
            for p in processes:
                p.start()
            for p in processes:
                p.join()

        for ID in l[0]:
            if not l[1].empty:
                cols_list = list(l[2][ID].columns)
                cols_list.remove('Type')
                cols_list[1:1] = ['Type']
                l[2][ID] = l[2][ID][cols_list]
                l[2][ID].update(new_dfs_dict[ID])
                l[2][ID] = pd.merge(l[2][ID], new_dfs_dict[ID])
            else:
                l[2][ID] = new_dfs_dict[ID]
        ready_to_set_df = pd.DataFrame()
        for ID in l[2]:
            ready_to_set_df = pd.concat([ready_to_set_df, l[2][ID]], sort=False)
            output_sheet = wb01.worksheet_by_title(l[3])            
            output_sheet.clear()
            output_sheet.set_dataframe(ready_to_set_df.fillna(0), 'A1')

    output_df = pd.DataFrame(columns=bank_portal_data_header)
    for df in output_dfs:
        output_df = output_df.append(df)
    try:
        data_sheet = wb02.worksheet_by_title(end_date)
        data_df = data_sheet.get_as_df(has_header=True)
        output_df = data_df[bank_portal_data_header].append(output_df)
        output_df = output_df.drop_duplicates(keep=False)
        data_sheet.clear()
        data_sheet.set_dataframe(output_df, 'A1', copy_head=True)
    except:
        data_sheet = wb02.add_worksheet(end_date)
        data_sheet.set_dataframe(output_df, 'A1', copy_head=True)`````

It first threw the exception at the line data_sheet.clear().
When I commented out this statement, it was at the next line data_sheet.set_dataframe(output_df, 'A1', copy_head=True)

Please help!

### PR DESCRIPTION
**Pull request by felipemaion**

### CODE REVIEW

**Comment by felipemaion** [comment]
@nithinmurali I was not able to write the code for the tests, since it was difficult to reproduce the error while multithreading (SSLError). It is working like a charm in my application and all tests passed.
The fix is based on the [Google APIs page regarding the threading problem with httplib2](https://googleapis.github.io/google-api-python-client/docs/thread_safety.html)

**Comment by felipemaion** [comment]
@nithinmurali another thing for the future due to deprecating (from pytest report):

```
=============================================================================================== warnings summary ================================================================================================
tests/authorization_test.py::TestAuthorization::test_deprecated_kwargs_removal
tests/authorization_test.py::TestAuthorization::test_kwargs_passed_to_client
 /pygsheets/pygsheets/authorization.py:129: DeprecationWarning: The argument service_file is deprecated. Use service_account_file instead.
    warnings.warn('The argument {} is deprecated. Use {} instead.'.format(key, _deprecated_keyword_mapping[key])

tests/online_test.py::TestClient::test_open_title
  /pygsheets/tests/online_test.py:32: DeprecationWarning: This method will be removed in Python 3.12. Use 'parser.read_file()' instead.
    config.readfp(open(filename))

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
================================================================================== 71 passed, 3 warnings in 200.91s (0:03:20) ===================================================================================
(
```

File: pygsheets/drive.py [hunk 1]
```diff
@@ -48,13 +48,13 @@ class DriveAPIWrapper(object):
     _spreadsheet_mime_type = "application/vnd.google-apps.spreadsheet"
     _folder_mime_type = "application/vnd.google-apps.folder"
 
-    def __init__(self, http, data_path, retries=3, logger=logging.getLogger(__name__)):
+    def __init__(self, http, data_path, retries=3, logger=logging.getLogger(__name__), requestBuilder=None):
```
Comment (nithinmurali) [inline pygsheets/drive.py]: please change `requestBuilder` to `request_builder` for consistency.

File: pygsheets/sheet.py [hunk 2]
```diff
@@ -18,7 +18,7 @@
 class SheetAPIWrapper(object):
 
     def __init__(self, http, data_path, seconds_per_quota=100, retries=1, logger=logging.getLogger(__name__),
-                 check=True):
+                 check=True, requestBuilder=None):
```
Comment (nithinmurali) [inline pygsheets/sheet.py]: please change `requestBuilder` to `request_builder` for consistency.

File: pygsheets/client.py [hunk 3]
```diff
@@ -55,12 +57,16 @@ def __init__(self, credentials, retries=3, http=None, check=True, seconds_per_qu
...
+        http = AuthorizedHttp(credentials, http=httplib2.Http())
         data_path = os.path.join(os.path.dirname(os.path.abspath(__file__)), "data")
 
-        self.sheet = SheetAPIWrapper(http, data_path, retries=retries, check=check, seconds_per_quota=seconds_per_quota)
-        self.drive = DriveAPIWrapper(http, data_path)
+        self.sheet = SheetAPIWrapper(http, data_path, retries=retries, check=check, seconds_per_quota=seconds_per_quota, requestBuilder=self.build_request)
+        self.drive = DriveAPIWrapper(http, data_path, requestBuilder=self.build_request)
 
+
+    def build_request(self, http, *args, **kwargs):
```
Comment (nithinmurali) [inline pygsheets/client.py]: this should be a private function

File: pygsheets/client.py [hunk 4]
```diff
@@ -55,12 +57,16 @@ def __init__(self, credentials, retries=3, http=None, check=True, seconds_per_qu
         self.oauth = credentials
         self.logger = logging.getLogger(__name__)
 
-        http = AuthorizedHttp(credentials, http=http)
+        http = AuthorizedHttp(credentials, http=httplib2.Http())
```
Comment (nithinmurali) [inline pygsheets/client.py]: this means we are not using the user-passed `http` param anymore, so it is a breaking change. i think this should be `http or httplib2.Http()`

**Review by nithinmurali — CHANGES_REQUESTED** [review]
Needs some minor styling fixes.

File: pygsheets/drive.py (hunk 1 above)
Comment (felipemaion) [reply pygsheets/drive.py]: Done.

File: pygsheets/sheet.py (hunk 2 above)
Comment (felipemaion) [reply pygsheets/sheet.py]: Done.

File: pygsheets/client.py (hunk 3 above)
Comment (felipemaion) [reply pygsheets/client.py]: __build_request.
Done

File: pygsheets/client.py (hunk 4 above)
Comment (felipemaion) [reply pygsheets/client.py]: I've created a conditional for that.

**Comment by felipemaion** [comment]
@nithinmurali please review the new push request. 
Thank you

**Comment by felipemaion** [comment]
@nithinmurali It was showing here tests error with Python 3.7. 
I just created an env with python=3.7 and run the pytest. Everything passed.

**Review by nithinmurali — APPROVED** [review]

**Comment by nithinmurali** [comment]
@felipemaion I think it should be fine. The online test can be flaky as it tests against real API. sometimes connection problems can arise.

**Pull request merged**